### ISSUE TITLE
Add `mem_gb` to SUSAN and MELODIC nodes to avoid memory errors

### ISSUE THREAD

**Issue by HippocampusGirl** [nipreps/fmriprep#2478]
I have been having some issues with `BrokenProcessPool` errors (i.e., out-of-memory errors), and these usually occur around the `melodic` and `susan` commands. Would it be possible to add a `mem_gb` of a few gigabytes to both of these, so that the `nipype` scheduler knows not to start too many at the same time?

https://github.com/nipreps/fmriprep/blob/a45ea486a8aa58541add3370a60ae87a70af8efb/fmriprep/workflows/bold/confounds.py#L668-L673

**Comment by effigies** [comment]
Absolutely. Do you have a sense of how much they use, as a function of input file size?

**Comment by HippocampusGirl** [comment]
I believe these nodes will only ever be called with resampled data in `MNI152NLin6Asym` space, so the only parameter would be the number of volumes. I do not have a sense of that though, I would need to set up a few test runs.

I think it may also be a challenge to estimate memory usage for `melodic`, because it infers the number of components it estimates automatically, and that number may also impact memory usage.

**Comment by effigies** [comment]
At a guess, 2x the uncompressed 4D array should be the dominant term in memory complexity. For SUSAN, the pre-smoothed and smoothed arrays. For MELODIC, the maximum number of components would be number of original volumes.

Might be another significant term I'm not thinking of, but we could try that to start.

**Comment by HippocampusGirl** [comment]
I just checked, one of our uncompressed inputs to `susan` was 1.6 GB in size, but `susan` used more than 10 GB of memory. I'm actually not sure what kind of memory complexity it has.

**Comment by effigies** [comment]
Can some of it be explained by a type conversion? If you have int16 data, that'll be promoted to float or double. 2 x prod(shape) x sizeof(double) might get pretty close.

**Comment by HippocampusGirl** [comment]
Yes, that seems to be exactly what is happening. I tested two different images, and the calculation you suggested seems to be accurate for the `susan` part, and even has a few gigabytes of safety margin. For `melodic`, it seems that 1 x prod(shape) x sizeof(double) would be sufficient.

**Comment by HippocampusGirl** [comment]
I would be happy to contribute a pull request, but I'm not sure how deal with the type conversion. As far as I can tell, the workflow only has the `mem_gb` argument, and that contains the input file size. I don't think the workflow function will be able to tell if the file size refers to int16 data that is going to be converted or a file that is already float64.

**Comment by effigies** [comment]
I wouldn't worry about the type conversion and just assume it's double. So I think we just need to know the xyz size of the MNI volume to come up with a scale factor that we then multiply by the number of volumes in the series.

**Comment by tsalo** [comment]
I think this can be transferred to `fmripost-aroma`.
